records/2021-08-16.scalability.move-api-to-lambda.md

```diff
@@ -0,0 +1,42 @@
+# Move the Notify API from Kubernetes to AWS Lambda
+
+Date: 2021-08-16
+
+## Status
+
+**PROPOSED**
+
+## Context
+
+The SRE team is encouraging us to move the Notify API from Kubernetes to AWS Lambda. AWS Lambda functions allow code to be run in a highly scalable way - we can have hundreds of lambda functions running simulateously, and this auto-scales to meet the number of requests being made to the API. The benefits of doing this are thought to be the following:
+
+- less complexity for the Notify team to manage, compared to Kubernetes
+- more scalable API
+- less costly compared to kubernetes + Amazon EKS
+
+At the time when Notify was being deployed by CDS, AWS Lambda had a smaller maximum size for a lambda function. Notify was over this limit and so @max deployed it a different way.
+
+Since then, AWS has added support for Docker images and increased the size limit for these to 10 GB. This means it is now possible to deploy the Notify API as a Lambda function.
+
+CDS has now used lambda functions to run the covid alert metrics collection and aggregation (https://github.com/cds-snc/covid-alert-metrics-terraform/tree/main/aws). Lamba functions helped that part of the service to scale well.
+
+## Proposed architecture
+
+AWS Lambda + API Gateway
+
+## Additional considerations
```

Comment:
Might want to consider the tradeoffs of local testing, I'm assuming that Kubernetes might be an easier solution to test locally since as long as you have docker desktop installed you can probably bring up the test system pretty easily. 

Lambdas make local testing a bit more difficult by forcing you to either wire up some docker containers using aws-lambda-rie or use AWS SAM to orchestrate the work.

Reply:
Thanks @CalvinRodo I added this in.

@jimleroyer Are you or anyone else doing local testing with Kubernetes?